### PR TITLE
Hardening: input validation and bounds tightening across 26 subsystems

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
General hardening pass tightening input validation, buffer bounds, and protocol handling across the runtime.

Areas touched (37 commits):

- node:path (1)
- node:buffer (1)
- node:net (1)
- node:http (2)
- node:child_process (1)
- http2 (1)
- http (1)
- fetch (1)
- websocket (3)
- Bun.serve (1)
- bake / dev server (1)
- sql / postgres (3)
- sql / mysql (1)
- sql (1)
- valkey (1)
- sqlite (1)
- install / lockfile / patch (4)
- resolver (1)
- archive (1)
- s3 (1)
- sys (1)
- glob (1)
- css (1)
- bundler (1)
- router (1)
- debugger (1)
- webcore / blob / serialization (2)
- webcrypto (1)

`cargo check -p bun_bin` and `bun run rust:check-all` (all 10 targets) pass.

### CODE REVIEW

File: src/jsc/bindings/JSBuffer.cpp [hunk 1]
```diff
@@ -2423,6 +2428,14 @@ static JSC::EncodedJSValue jsBufferPrototypeFunction_writeBody(JSC::JSGlobalObje
     auto encoding = parseEncoding(scope, lexicalGlobalObject, encodingValue, false);
     RETURN_IF_EXCEPTION(scope, {});
 
+    if (castedThis->isDetached()) [[unlikely]] {
+        throwTypeError(lexicalGlobalObject, scope, "ArrayBufferView is detached"_s);
+        return {};
+    }
+    uint32_t currentByteLength = castedThis->byteLength();
+    uint32_t currentRemaining = offset < currentByteLength ? currentByteLength - offset : 0;
+    if (length > currentRemaining) length = currentRemaining;
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/jsc/bindings/JSBuffer.cpp:2439]: `offset` is bounded by `castedThis->byteLength()` at JSBuffer.cpp:2402 before any re-entrant call, and the backing allocation never relocates or shrinks afterward — JSC resizable ArrayBuffers reserve `maxByteLength` up front, and non-resizable buffers can only detach (caught at :2431). So `vector() + offset` remains inside the original allocation even if the logical `byteLength()` drops below `offset`; with `length` clamped to 0 at :2437 the encoders at :419/:422 perform no store. No out-of-range write is reachable.

File: src/libarchive/lib.rs [hunk 2]
```diff
@@ -1331,7 +1346,14 @@ fn path_traverses_created_symlink(path: &[u8], created_symlinks: &[Vec<u8>]) ->
...
             let prefix = &path[..end];
-            if !prefix.is_empty() && created_symlinks.iter().any(|s| s.as_slice() == prefix) {
+            // Compare case-insensitively: on case-insensitive filesystems (APFS
+            // default on macOS) an entry `LINK/x` traverses a symlink stored as
+            // `link`, but a byte-exact compare would miss it.
+            if !prefix.is_empty()
+                && created_symlinks
+                    .iter()
+                    .any(|s| strings::eql_case_insensitive_ascii_check_length(s, prefix))
+            {
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/libarchive/lib.rs:1383]: Intentional. Gating on FS case-sensitivity would require a runtime probe (no helper exists in the tree — see grep for `case_insensitive_fs`/`casefold`), and a `cfg(target_os)` gate would re-open the traversal hole on Linux ext4-casefold / mounted APFS/exFAT, which is exactly what this hardening closes. The only false-positive is an archive that ships both `link` (symlink) and `LINK/...` as distinct siblings; such archives are already non-portable to macOS/Windows and the consequence is just a skipped entry with a warning (src/libarchive/lib.rs:1830-1838). Keeping the comparison unconditionally case-insensitive is the conservative choice and narrowing it would weaken the original fix.

File: src/runtime/node/net/BlockList.rs [hunk 3]
```diff
@@ -33,7 +33,14 @@ use core::sync::atomic::{AtomicU32, Ordering as AtomicOrdering};
...
 use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsCell, JsResult, StringJsc as _};
-use bun_threading::Mutex;
+use bun_threading::{Guarded, Mutex};
+
+/// Addresses of `BlockList` instances currently embedded in a live
+/// `SerializedScriptValue` (one entry per serialize; removed by
+/// `BlockList__onStructuredCloneDestroy`). Deserialize only honours pointers
+/// present here so wire bytes from another process (IPC `advanced` mode,
+/// `node:v8.deserialize`) cannot smuggle an arbitrary address through tag 251.
+static SERIALIZED_REFS: Guarded<Vec<usize>> = Guarded::new(Vec::new());
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/runtime/node/net/BlockList.rs:43]: Deferring to a follow-up. The current registry already preserves memory safety: each `push(addr)` is paired with `ref_()` and each `swap_remove(addr)` with `deref()`, so any address present in `SERIALIZED_REFS` is guaranteed live — allocator reuse cannot land a different object there while it's registered. Moving to a per-serialization token additionally requires changing the `BlockList__onStructuredCloneDestroy(*mut c_void)` C++ ABI so the destructor can hand the token back, which is outside the scope of this sweep.

File: src/runtime/webcore/Blob.rs [hunk 4]
```diff
@@ -2664,9 +2664,10 @@ impl BlobExt for Blob {
...
             let _free = (LIFETIME == Lifetime::Temporary).then(|| TemporaryBytes(raw_bytes));
-            // BOM::Utf16Le ⇒ buf is UTF-16LE bytes; len is even after BOM strip.
-            // Mirrors Zig `bun.reinterpretSlice(u16, buf)`; bytemuck checks align + even-len.
+            // Mirrors Zig `bun.reinterpretSlice(u16, buf)`: drop a trailing odd byte
+            // (`@divTrunc`) so `bytemuck::cast_slice` cannot panic on slop.
             // +1 WTF ref; `OwnedString` releases it on scope exit (Zig: `defer out.deref()`).
+            let buf = &buf[..buf.len() & !1];
             let out =
                 OwnedString::new(BunString::clone_utf16(bytemuck::cast_slice::<u8, u16>(buf)));
             return out.to_js(global);
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/runtime/webcore/Blob.rs:2723]: This PR's only change here is the odd-byte truncation (`buf.len() & !1`); the Transfer/detach behavior is untouched from `main` and mirrors the Zig reference exactly — `src/runtime/webcore/Blob.zig:3775-3780` does **not** call `this.detach()` on the UTF-16LE branch under `.transfer` (the asymmetry with `toJSONWithBytes` at `Blob.zig:3892` is upstream). Per repo policy the `.zig` sibling is the semantics source of truth, so diverging from it in a hardening sweep is out of scope; if the asymmetry is a real bug it should be filed against the Zig original separately.

File: src/js/node/child_process.ts [hunk 5]
```diff
@@ -952,6 +952,11 @@ function normalizeSpawnArguments(file, args, options) {
       else file = "/bin/sh";
       args = ["-c", command];
     }
+  } else if (process.platform === "win32" && /\.(?:cmd|bat)[\s.]*$/i.exec(file) !== null) {
+    // CVE-2024-27980: CreateProcess routes .bat/.cmd through cmd.exe, which
+    // re-parses argv with shell metacharacter rules. Reject unless the caller
+    // explicitly opted into shell semantics.
+    throw new SystemError(`spawn EINVAL: ${file}`, file, "spawn", -4071, "EINVAL");
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/js/node/child_process.ts]: Addressed in c48d2dc60928 — `normalizeSpawnArguments` now surfaces the rejection via the returned options instead of throwing; `spawn()` re-throws synchronously (matching the Node test's try/catch path) and `spawnSync()` returns it as `result.error` with `syscall`/`path`/`spawnargs` set.

File: src/runtime/api/bun/h2_frame_parser.rs [hunk 6]
```diff
@@ -4028,34 +4037,46 @@ impl H2FrameParser {
...
+            .get()
+            .unwrap_or_else(|| self.local_settings.get());
+        if frame.length > settings.max_frame_size {
+            self.send_go_away(
+                frame.stream_identifier,
+                ErrorCode::FRAME_SIZE_ERROR,
+                b"invalid Continuation frame size",
+                self.last_stream_id.get(),
+                true,
+            );
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/runtime/api/bun/h2_frame_parser.rs]: Already addressed in bc270dcd0bff (same push) — `handle_continuation_frame` now reads `self.local_settings.get().max_frame_size` exactly like the sibling handlers; the quoted `remote_settings.get().unwrap_or_else(...)` line no longer exists in the file (verified: `rg remote_settings.get..unwrap_or_else` returns no matches).

File: src/jsc/bindings/JSBuffer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/jsc/bindings/JSBuffer.cpp:2439]: Addressed in 2e7b241956b2.

File: src/runtime/node/net/BlockList.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/runtime/node/net/BlockList.rs:43]: Addressed in e26e72c24841.

File: src/runtime/webcore/Blob.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/runtime/webcore/Blob.rs:2723]: Addressed in 2529459224fe.

File: src/runtime/bake/DevServer.rs [hunk 7]
```diff
@@ -1540,6 +1601,9 @@ impl<const SSL: bool> bun_uws_sys::web_socket::WebSocketUpgradeServer<SSL> for D
         // SAFETY: uWS guarantees `res` is non-null and live for the upgrade
         // callback; `Response<SSL>` is an opaque handle.
         let res = unsafe { &mut *res };
+        if !is_allowed_dev_host(this, req) {
+            return host_forbidden(res.as_any_response());
+        }
```
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/runtime/bake/DevServer.rs:1599]: Addressed in 988e969aab74.

File: src/libarchive/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jarred-Sumner) [reply src/libarchive/lib.rs:1383]: Not applying — gating the case-insensitive comparison would weaken the original hardening on case-insensitive Linux mounts (ext4-casefold, exFAT/APFS), per the accepted refutation above. The unconditional comparison stays.
